[SymmetricMemory] make sure different subgroups with the same name use different store prefixes

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130842
* #130757
* __->__ #130756

This fixes a race condition in which different subgroups with the same name on the same host would use the same store.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

Commit:
[SymmetricMemory] make sure different subgroups with the same name use different store prefixes

[ghstack-poisoned]

Skylion007 commented on torch/distributed/_symmetric_memory/__init__.py:
```diff
@@ -24,10 +24,14 @@ def enable_symm_mem_for_group(group_name: str) -> None:
     group = c10d._resolve_process_group(group_name)
+    global_ranks = sorted(c10d._world.pg_group_ranks[group].keys())
+    # Different subgroups with the same name should use different stores
+    global_ranks_str = "_".join(str(r) for r in global_ranks)
```
```suggestion
    global_ranks_str = "_".join(map(str, global_ranks))
```

wanchaol commented on torch/distributed/_symmetric_memory/__init__.py:
```diff
@@ -24,10 +24,14 @@ def enable_symm_mem_for_group(group_name: str) -> None:
+    global_ranks_str = "_".join(str(r) for r in global_ranks)
     store = c10d.PrefixStore(
-        "symmetric_memory",
+        f"symmetric_memory-{global_ranks_str}",
```
curious can't we use f"symmetric_memory-{group_name}" directly?

iirc `group_name` is unique already for all subgroups

yifuwang replied:
Oh are they? I'll check it out. Your suggestion would be better if that's the case

Commit:
Update on "[SymmetricMemory] make sure different subgroups with the same name use different store prefixes"


This fixes a race condition in which different subgroups with the same name on the same host would use the same store.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]
